File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/EnvironmentalPhysicalLeg.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class EnvironmentalPhysicalLeg : PhysicalSwapLeg
    {
        public EnvironmentalPhysicalLeg(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode numberOfAllowancesNode = xmlNode.SelectSingleNode("numberOfAllowances");
            
            if (numberOfAllowancesNode != null)
            {
                if (numberOfAllowancesNode.Attributes["href"] != null || numberOfAllowancesNode.Attributes["id"] != null) 
                {
                    if (numberOfAllowancesNode.Attributes["id"] != null) 
                    {
                        numberOfAllowancesIDRef_ = numberOfAllowancesNode.Attributes["id"].Value;
                        UnitQuantity ob = new UnitQuantity(numberOfAllowancesNode);
                        IDManager.SetID(numberOfAllowancesIDRef_, ob);
                    }
                    else if (numberOfAllowancesNode.Attributes["href"] != null)
                    {
                        numberOfAllowancesIDRef_ = numberOfAllowancesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        numberOfAllowances_ = new UnitQuantity(numberOfAllowancesNode);
                    }
                }
                else
                {
                    numberOfAllowances_ = new UnitQuantity(numberOfAllowancesNode);
                }
            }
            
        
            XmlNode environmentalNode = xmlNode.SelectSingleNode("environmental");
            
            if (environmentalNode != null)
            {
                if (environmentalNode.Attributes["href"] != null || environmentalNode.Attributes["id"] != null) 
                {
                    if (environmentalNode.Attributes["id"] != null) 
                    {
                        environmentalIDRef_ = environmentalNode.Attributes["id"].Value;
                        EnvironmentalProduct ob = new EnvironmentalProduct(environmentalNode);
                        IDManager.SetID(environmentalIDRef_, ob);
                    }
                    else if (environmentalNode.Attributes["href"] != null)
                    {
                        environmentalIDRef_ = environmentalNode.Attributes["href"].Value;
                    }
                    else
                    {
                        environmental_ = new EnvironmentalProduct(environmentalNode);
                    }
                }
                else
                {
                    environmental_ = new EnvironmentalProduct(environmentalNode);
                }
            }
            
        
            XmlNode abandonmentOfSchemeNode = xmlNode.SelectSingleNode("abandonmentOfScheme");
            
            if (abandonmentOfSchemeNode != null)
            {
                if (abandonmentOfSchemeNode.Attributes["href"] != null || abandonmentOfSchemeNode.Attributes["id"] != null) 
                {
                    if (abandonmentOfSchemeNode.Attributes["id"] != null) 
                    {
                        abandonmentOfSchemeIDRef_ = abandonmentOfSchemeNode.Attributes["id"].Value;
                        EnvironmentalAbandonmentOfSchemeEnum ob = new EnvironmentalAbandonmentOfSchemeEnum(abandonmentOfSchemeNode);
                        IDManager.SetID(abandonmentOfSchemeIDRef_, ob);
                    }
                    else if (abandonmentOfSchemeNode.Attributes["href"] != null)
                    {
                        abandonmentOfSchemeIDRef_ = abandonmentOfSchemeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        abandonmentOfScheme_ = new EnvironmentalAbandonmentOfSchemeEnum(abandonmentOfSchemeNode);
                    }
                }
                else
                {
                    abandonmentOfScheme_ = new EnvironmentalAbandonmentOfSchemeEnum(abandonmentOfSchemeNode);
                }
            }
            
        
            XmlNode deliveryDateNode = xmlNode.SelectSingleNode("deliveryDate");
            
            if (deliveryDateNode != null)
            {
                if (deliveryDateNode.Attributes["href"] != null || deliveryDateNode.Attributes["id"] != null) 
                {
                    if (deliveryDateNode.Attributes["id"] != null) 
                    {
                        deliveryDateIDRef_ = deliveryDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(deliveryDateNode);
                        IDManager.SetID(deliveryDateIDRef_, ob);
                    }
                    else if (deliveryDateNode.Attributes["href"] != null)
                    {
                        deliveryDateIDRef_ = deliveryDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        deliveryDate_ = new AdjustableOrRelativeDate(deliveryDateNode);
                    }
                }
                else
                {
                    deliveryDate_ = new AdjustableOrRelativeDate(deliveryDateNode);
                }
            }
            
        
            XmlNode paymentDateNode = xmlNode.SelectSingleNode("paymentDate");
            
            if (paymentDateNode != null)
            {
                if (paymentDateNode.Attributes["href"] != null || paymentDateNode.Attributes["id"] != null) 
                {
                    if (paymentDateNode.Attributes["id"] != null) 
                    {
                        paymentDateIDRef_ = paymentDateNode.Attributes["id"].Value;
                        DateOffset ob = new DateOffset(paymentDateNode);
                        IDManager.SetID(paymentDateIDRef_, ob);
                    }
                    else if (paymentDateNode.Attributes["href"] != null)
                    {
                        paymentDateIDRef_ = paymentDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        paymentDate_ = new DateOffset(paymentDateNode);
                    }
                }
                else
                {
                    paymentDate_ = new DateOffset(paymentDateNode);
                }
            }
            
        
            XmlNode businessCentersReferenceNode = xmlNode.SelectSingleNode("businessCentersReference");
            
            if (businessCentersReferenceNode != null)
            {
                if (businessCentersReferenceNode.Attributes["href"] != null || businessCentersReferenceNode.Attributes["id"] != null) 
                {
                    if (businessCentersReferenceNode.Attributes["id"] != null) 
                    {
                        businessCentersReferenceIDRef_ = businessCentersReferenceNode.Attributes["id"].Value;
                        BusinessCentersReference ob = new BusinessCentersReference(businessCentersReferenceNode);
                        IDManager.SetID(businessCentersReferenceIDRef_, ob);
                    }
                    else if (businessCentersReferenceNode.Attributes["href"] != null)
                    {
                        businessCentersReferenceIDRef_ = businessCentersReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        businessCentersReference_ = new BusinessCentersReference(businessCentersReferenceNode);
                    }
                }
                else
                {
                    businessCentersReference_ = new BusinessCentersReference(businessCentersReferenceNode);
                }
            }
            
        
            XmlNode businessCentersNode = xmlNode.SelectSingleNode("businessCenters");
            
            if (businessCentersNode != null)
            {
                if (businessCentersNode.Attributes["href"] != null || businessCentersNode.Attributes["id"] != null) 
                {
                    if (businessCentersNode.Attributes["id"] != null) 
                    {
                        businessCentersIDRef_ = businessCentersNode.Attributes["id"].Value;
                        BusinessCenters ob = new BusinessCenters(businessCentersNode);
                        IDManager.SetID(businessCentersIDRef_, ob);
                    }
                    else if (businessCentersNode.Attributes["href"] != null)
                    {
                        businessCentersIDRef_ = businessCentersNode.Attributes["href"].Value;
                    }
                    else
                    {
                        businessCenters_ = new BusinessCenters(businessCentersNode);
                    }
                }
                else
                {
                    businessCenters_ = new BusinessCenters(businessCentersNode);
                }
            }
            
        
            XmlNode failureToDeliverApplicableNode = xmlNode.SelectSingleNode("failureToDeliverApplicable");
            
            if (failureToDeliverApplicableNode != null)
            {
                if (failureToDeliverApplicableNode.Attributes["href"] != null || failureToDeliverApplicableNode.Attributes["id"] != null) 
                {
                    if (failureToDeliverApplicableNode.Attributes["id"] != null) 
                    {
                        failureToDeliverApplicableIDRef_ = failureToDeliverApplicableNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(failureToDeliverApplicableNode);
                        IDManager.SetID(failureToDeliverApplicableIDRef_, ob);
                    }
                    else if (failureToDeliverApplicableNode.Attributes["href"] != null)
                    {
                        failureToDeliverApplicableIDRef_ = failureToDeliverApplicableNode.Attributes["href"].Value;
                    }
                    else
                    {
                        failureToDeliverApplicable_ = new XsdTypeBoolean(failureToDeliverApplicableNode);
                    }
                }
                else
                {
                    failureToDeliverApplicable_ = new XsdTypeBoolean(failureToDeliverApplicableNode);
                }
            }
            
        
            XmlNode eEPParametersNode = xmlNode.SelectSingleNode("eEPParameters");
            
            if (eEPParametersNode != null)
            {
                if (eEPParametersNode.Attributes["href"] != null || eEPParametersNode.Attributes["id"] != null) 
                {
                    if (eEPParametersNode.Attributes["id"] != null) 
                    {
                        eEPParametersIDRef_ = eEPParametersNode.Attributes["id"].Value;
                        EEPParameters ob = new EEPParameters(eEPParametersNode);
                        IDManager.SetID(eEPParametersIDRef_, ob);
                    }
                    else if (eEPParametersNode.Attributes["href"] != null)
                    {
                        eEPParametersIDRef_ = eEPParametersNode.Attributes["href"].Value;
                    }
                    else
                    {
                        eEPParameters_ = new EEPParameters(eEPParametersNode);
                    }
                }
                else
                {
                    eEPParameters_ = new EEPParameters(eEPParametersNode);
                }
            }
            
        
        }
        
    
        #region NumberOfAllowances_
        private UnitQuantity numberOfAllowances_;
        public UnitQuantity NumberOfAllowances_
        {
            get
            {
                if (this.numberOfAllowances_ != null)
                {
                    return this.numberOfAllowances_; 
                }
                else if (this.numberOfAllowancesIDRef_ != null)
                {
                    numberOfAllowances_ = IDManager.getID(numberOfAllowancesIDRef_) as UnitQuantity;
                    return this.numberOfAllowances_; 
                }
                else
                {
                      return this.numberOfAllowances_; 
                }
            }
            set
            {
                if (this.numberOfAllowances_ != value)
                {
                    this.numberOfAllowances_ = value;
                }
            }
        }
        #endregion
        
        public string numberOfAllowancesIDRef_ { get; set; }
        #region Environmental_
        private EnvironmentalProduct environmental_;
        public EnvironmentalProduct Environmental_
        {
            get
            {
                if (this.environmental_ != null)
                {
                    return this.environmental_; 
                }
                else if (this.environmentalIDRef_ != null)
                {
                    environmental_ = IDManager.getID(environmentalIDRef_) as EnvironmentalProduct;
                    return this.environmental_; 
                }
                else
                {
                      return this.environmental_; 
                }
            }
            set
            {
                if (this.environmental_ != value)
                {
                    this.environmental_ = value;
                }
            }
        }
        #endregion
        
        public string environmentalIDRef_ { get; set; }
        #region AbandonmentOfScheme_
        private EnvironmentalAbandonmentOfSchemeEnum abandonmentOfScheme_;
        public EnvironmentalAbandonmentOfSchemeEnum AbandonmentOfScheme_
        {
            get
            {
                if (this.abandonmentOfScheme_ != null)
                {
                    return this.abandonmentOfScheme_; 
                }
                else if (this.abandonmentOfSchemeIDRef_ != null)
                {
                    abandonmentOfScheme_ = IDManager.getID(abandonmentOfSchemeIDRef_) as EnvironmentalAbandonmentOfSchemeEnum;
                    return this.abandonmentOfScheme_; 
                }
                else
                {
                      return this.abandonmentOfScheme_; 
                }
            }
            set
            {
                if (this.abandonmentOfScheme_ != value)
                {
                    this.abandonmentOfScheme_ = value;
                }
            }
        }
        #endregion
        
        public string abandonmentOfSchemeIDRef_ { get; set; }
        #region DeliveryDate_
        private AdjustableOrRelativeDate deliveryDate_;
        public AdjustableOrRelativeDate DeliveryDate_
        {
            get
            {
                if (this.deliveryDate_ != null)
                {
                    return this.deliveryDate_; 
                }
                else if (this.deliveryDateIDRef_ != null)
                {
                    deliveryDate_ = IDManager.getID(deliveryDateIDRef_) as AdjustableOrRelativeDate;
                    return this.deliveryDate_; 
                }
                else
                {
                      return this.deliveryDate_; 
                }
            }
            set
            {
                if (this.deliveryDate_ != value)
                {
                    this.deliveryDate_ = value;
                }
            }
        }
        #endregion
        
        public string deliveryDateIDRef_ { get; set; }
        #region PaymentDate_
        private DateOffset paymentDate_;
        public DateOffset PaymentDate_
        {
            get
            {
                if (this.paymentDate_ != null)
                {
                    return this.paymentDate_; 
                }
                else if (this.paymentDateIDRef_ != null)
                {
                    paymentDate_ = IDManager.getID(paymentDateIDRef_) as DateOffset;
                    return this.paymentDate_; 
                }
                else
                {
                      return this.paymentDate_; 
                }
            }
            set
            {
                if (this.paymentDate_ != value)
                {
                    this.paymentDate_ = value;
                }
            }
        }
        #endregion
        
        public string paymentDateIDRef_ { get; set; }
        #region BusinessCentersReference_
        private BusinessCentersReference businessCentersReference_;
        public BusinessCentersReference BusinessCentersReference_
        {
            get
            {
                if (this.businessCentersReference_ != null)
                {
                    return this.businessCentersReference_; 
                }
                else if (this.businessCentersReferenceIDRef_ != null)
                {
                    businessCentersReference_ = IDManager.getID(businessCentersReferenceIDRef_) as BusinessCentersReference;
                    return this.businessCentersReference_; 
                }
                else
                {
                      return this.businessCentersReference_; 
                }
            }
            set
            {
                if (this.businessCentersReference_ != value)
                {
                    this.businessCentersReference_ = value;
                }
            }
        }
        #endregion
        
        public string businessCentersReferenceIDRef_ { get; set; }
        #region BusinessCenters_
        private BusinessCenters businessCenters_;
        public BusinessCenters BusinessCenters_
        {
            get
            {
                if (this.businessCenters_ != null)
                {
                    return this.businessCenters_; 
                }
                else if (this.businessCentersIDRef_ != null)
                {
                    businessCenters_ = IDManager.getID(businessCentersIDRef_) as BusinessCenters;
                    return this.businessCenters_; 
                }
                else
                {
                      return this.businessCenters_; 
                }
            }
            set
            {
                if (this.businessCenters_ != value)
                {
                    this.businessCenters_ = value;
                }
            }
        }
        #endregion
        
        public string businessCentersIDRef_ { get; set; }
        #region FailureToDeliverApplicable_
        private XsdTypeBoolean failureToDeliverApplicable_;
        public XsdTypeBoolean FailureToDeliverApplicable_
        {
            get
            {
                if (this.failureToDeliverApplicable_ != null)
                {
                    return this.failureToDeliverApplicable_; 
                }
                else if (this.failureToDeliverApplicableIDRef_ != null)
                {
                    failureToDeliverApplicable_ = IDManager.getID(failureToDeliverApplicableIDRef_) as XsdTypeBoolean;
                    return this.failureToDeliverApplicable_; 
                }
                else
                {
                      return this.failureToDeliverApplicable_; 
                }
            }
            set
            {
                if (this.failureToDeliverApplicable_ != value)
                {
                    this.failureToDeliverApplicable_ = value;
                }
            }
        }
        #endregion
        
        public string failureToDeliverApplicableIDRef_ { get; set; }
        #region EEPParameters_
        private EEPParameters eEPParameters_;
        public EEPParameters EEPParameters_
        {
            get
            {
                if (this.eEPParameters_ != null)
                {
                    return this.eEPParameters_; 
                }
                else if (this.eEPParametersIDRef_ != null)
                {
                    eEPParameters_ = IDManager.getID(eEPParametersIDRef_) as EEPParameters;
                    return this.eEPParameters_; 
                }
                else
                {
                      return this.eEPParameters_; 
                }
            }
            set
            {
                if (this.eEPParameters_ != value)
                {
                    this.eEPParameters_ = value;
                }
            }
        }
        #endregion
        
        public string eEPParametersIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/EnvironmentalPhysicalLeg.cpp
// EnvironmentalPhysicalLeg.cpp 
#include "EnvironmentalPhysicalLeg.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

EnvironmentalPhysicalLeg::EnvironmentalPhysicalLeg(TiXmlNode* xmlNode)
: PhysicalSwapLeg(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //numberOfAllowancesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* numberOfAllowancesNode = xmlNode->FirstChildElement("numberOfAllowances");

   if(numberOfAllowancesNode){numberOfAllowancesIsNull_ = false;}
   else{numberOfAllowancesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- numberOfAllowancesNode , address : " << numberOfAllowancesNode << std::endl;
   #endif
   if(numberOfAllowancesNode)
   {
      if(numberOfAllowancesNode->Attribute("href") || numberOfAllowancesNode->Attribute("id"))
      {
          if(numberOfAllowancesNode->Attribute("id"))
          {
             numberOfAllowancesIDRef_ = numberOfAllowancesNode->Attribute("id");
             numberOfAllowances_ = boost::shared_ptr<UnitQuantity>(new UnitQuantity(numberOfAllowancesNode));
             numberOfAllowances_->setID(numberOfAllowancesIDRef_);
             IDManager::instance().setID(numberOfAllowancesIDRef_,numberOfAllowances_);
          }
          else if(numberOfAllowancesNode->Attribute("href")) { numberOfAllowancesIDRef_ = numberOfAllowancesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { numberOfAllowances_ = boost::shared_ptr<UnitQuantity>(new UnitQuantity(numberOfAllowancesNode));}
   }

   //environmentalNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* environmentalNode = xmlNode->FirstChildElement("environmental");

   if(environmentalNode){environmentalIsNull_ = false;}
   else{environmentalIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- environmentalNode , address : " << environmentalNode << std::endl;
   #endif
   if(environmentalNode)
   {
      if(environmentalNode->Attribute("href") || environmentalNode->Attribute("id"))
      {
          if(environmentalNode->Attribute("id"))
          {
             environmentalIDRef_ = environmentalNode->Attribute("id");
             environmental_ = boost::shared_ptr<EnvironmentalProduct>(new EnvironmentalProduct(environmentalNode));
             environmental_->setID(environmentalIDRef_);
             IDManager::instance().setID(environmentalIDRef_,environmental_);
          }
          else if(environmentalNode->Attribute("href")) { environmentalIDRef_ = environmentalNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { environmental_ = boost::shared_ptr<EnvironmentalProduct>(new EnvironmentalProduct(environmentalNode));}
   }

   //abandonmentOfSchemeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* abandonmentOfSchemeNode = xmlNode->FirstChildElement("abandonmentOfScheme");

   if(abandonmentOfSchemeNode){abandonmentOfSchemeIsNull_ = false;}
   else{abandonmentOfSchemeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- abandonmentOfSchemeNode , address : " << abandonmentOfSchemeNode << std::endl;
   #endif
   if(abandonmentOfSchemeNode)
   {
      if(abandonmentOfSchemeNode->Attribute("href") || abandonmentOfSchemeNode->Attribute("id"))
      {
          if(abandonmentOfSchemeNode->Attribute("id"))
          {
             abandonmentOfSchemeIDRef_ = abandonmentOfSchemeNode->Attribute("id");
             abandonmentOfScheme_ = boost::shared_ptr<EnvironmentalAbandonmentOfSchemeEnum>(new EnvironmentalAbandonmentOfSchemeEnum(abandonmentOfSchemeNode));
             abandonmentOfScheme_->setID(abandonmentOfSchemeIDRef_);
             IDManager::instance().setID(abandonmentOfSchemeIDRef_,abandonmentOfScheme_);
          }
          else if(abandonmentOfSchemeNode->Attribute("href")) { abandonmentOfSchemeIDRef_ = abandonmentOfSchemeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { abandonmentOfScheme_ = boost::shared_ptr<EnvironmentalAbandonmentOfSchemeEnum>(new EnvironmentalAbandonmentOfSchemeEnum(abandonmentOfSchemeNode));}
   }

   //deliveryDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* deliveryDateNode = xmlNode->FirstChildElement("deliveryDate");

   if(deliveryDateNode){deliveryDateIsNull_ = false;}
   else{deliveryDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- deliveryDateNode , address : " << deliveryDateNode << std::endl;
   #endif
   if(deliveryDateNode)
   {
      if(deliveryDateNode->Attribute("href") || deliveryDateNode->Attribute("id"))
      {
          if(deliveryDateNode->Attribute("id"))
          {
             deliveryDateIDRef_ = deliveryDateNode->Attribute("id");
             deliveryDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(deliveryDateNode));
             deliveryDate_->setID(deliveryDateIDRef_);
             IDManager::instance().setID(deliveryDateIDRef_,deliveryDate_);
          }
          else if(deliveryDateNode->Attribute("href")) { deliveryDateIDRef_ = deliveryDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { deliveryDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(deliveryDateNode));}
   }

   //paymentDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* paymentDateNode = xmlNode->FirstChildElement("paymentDate");

   if(paymentDateNode){paymentDateIsNull_ = false;}
   else{paymentDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- paymentDateNode , address : " << paymentDateNode << std::endl;
   #endif
   if(paymentDateNode)
   {
      if(paymentDateNode->Attribute("href") || paymentDateNode->Attribute("id"))
      {
          if(paymentDateNode->Attribute("id"))
          {
             paymentDateIDRef_ = paymentDateNode->Attribute("id");
             paymentDate_ = boost::shared_ptr<DateOffset>(new DateOffset(paymentDateNode));
             paymentDate_->setID(paymentDateIDRef_);
             IDManager::instance().setID(paymentDateIDRef_,paymentDate_);
          }
          else if(paymentDateNode->Attribute("href")) { paymentDateIDRef_ = paymentDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { paymentDate_ = boost::shared_ptr<DateOffset>(new DateOffset(paymentDateNode));}
   }

   //businessCentersReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* businessCentersReferenceNode = xmlNode->FirstChildElement("businessCentersReference");

   if(businessCentersReferenceNode){businessCentersReferenceIsNull_ = false;}
   else{businessCentersReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- businessCentersReferenceNode , address : " << businessCentersReferenceNode << std::endl;
   #endif
   if(businessCentersReferenceNode)
   {
      if(businessCentersReferenceNode->Attribute("href") || businessCentersReferenceNode->Attribute("id"))
      {
          if(businessCentersReferenceNode->Attribute("id"))
          {
             businessCentersReferenceIDRef_ = businessCentersReferenceNode->Attribute("id");
             businessCentersReference_ = boost::shared_ptr<BusinessCentersReference>(new BusinessCentersReference(businessCentersReferenceNode));
             businessCentersReference_->setID(businessCentersReferenceIDRef_);
             IDManager::instance().setID(businessCentersReferenceIDRef_,businessCentersReference_);
          }
          else if(businessCentersReferenceNode->Attribute("href")) { businessCentersReferenceIDRef_ = businessCentersReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { businessCentersReference_ = boost::shared_ptr<BusinessCentersReference>(new BusinessCentersReference(businessCentersReferenceNode));}
   }

   //businessCentersNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* businessCentersNode = xmlNode->FirstChildElement("businessCenters");

   if(businessCentersNode){businessCentersIsNull_ = false;}
   else{businessCentersIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- businessCentersNode , address : " << businessCentersNode << std::endl;
   #endif
   if(businessCentersNode)
   {
      if(businessCentersNode->Attribute("href") || businessCentersNode->Attribute("id"))
      {
          if(businessCentersNode->Attribute("id"))
          {
             businessCentersIDRef_ = businessCentersNode->Attribute("id");
             businessCenters_ = boost::shared_ptr<BusinessCenters>(new BusinessCenters(businessCentersNode));
             businessCenters_->setID(businessCentersIDRef_);
             IDManager::instance().setID(businessCentersIDRef_,businessCenters_);
          }
          else if(businessCentersNode->Attribute("href")) { businessCentersIDRef_ = businessCentersNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { businessCenters_ = boost::shared_ptr<BusinessCenters>(new BusinessCenters(businessCentersNode));}
   }

   //failureToDeliverApplicableNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* failureToDeliverApplicableNode = xmlNode->FirstChildElement("failureToDeliverApplicable");

   if(failureToDeliverApplicableNode){failureToDeliverApplicableIsNull_ = false;}
   else{failureToDeliverApplicableIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- failureToDeliverApplicableNode , address : " << failureToDeliverApplicableNode << std::endl;
   #endif
   if(failureToDeliverApplicableNode)
   {
      if(failureToDeliverApplicableNode->Attribute("href") || failureToDeliverApplicableNode->Attribute("id"))
      {
          if(failureToDeliverApplicableNode->Attribute("id"))
          {
             failureToDeliverApplicableIDRef_ = failureToDeliverApplicableNode->Attribute("id");
             failureToDeliverApplicable_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(failureToDeliverApplicableNode));
             failureToDeliverApplicable_->setID(failureToDeliverApplicableIDRef_);
             IDManager::instance().setID(failureToDeliverApplicableIDRef_,failureToDeliverApplicable_);
          }
          else if(failureToDeliverApplicableNode->Attribute("href")) { failureToDeliverApplicableIDRef_ = failureToDeliverApplicableNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { failureToDeliverApplicable_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(failureToDeliverApplicableNode));}
   }

   //eEPParametersNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* eEPParametersNode = xmlNode->FirstChildElement("eEPParameters");

   if(eEPParametersNode){eEPParametersIsNull_ = false;}
   else{eEPParametersIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- eEPParametersNode , address : " << eEPParametersNode << std::endl;
   #endif
   if(eEPParametersNode)
   {
      if(eEPParametersNode->Attribute("href") || eEPParametersNode->Attribute("id"))
      {
          if(eEPParametersNode->Attribute("id"))
          {
             eEPParametersIDRef_ = eEPParametersNode->Attribute("id");
             eEPParameters_ = boost::shared_ptr<EEPParameters>(new EEPParameters(eEPParametersNode));
             eEPParameters_->setID(eEPParametersIDRef_);
             IDManager::instance().setID(eEPParametersIDRef_,eEPParameters_);
          }
          else if(eEPParametersNode->Attribute("href")) { eEPParametersIDRef_ = eEPParametersNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { eEPParameters_ = boost::shared_ptr<EEPParameters>(new EEPParameters(eEPParametersNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<UnitQuantity> EnvironmentalPhysicalLeg::getNumberOfAllowances()
{
   if(!this->numberOfAllowancesIsNull_){
        if(numberOfAllowancesIDRef_ != ""){
             return boost::shared_static_cast<UnitQuantity>(IDManager::instance().getID(numberOfAllowancesIDRef_));
        }else{
             return this->numberOfAllowances_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<UnitQuantity>();
   }
}
boost::shared_ptr<EnvironmentalProduct> EnvironmentalPhysicalLeg::getEnvironmental()
{
   if(!this->environmentalIsNull_){
        if(environmentalIDRef_ != ""){
             return boost::shared_static_cast<EnvironmentalProduct>(IDManager::instance().getID(environmentalIDRef_));
        }else{
             return this->environmental_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EnvironmentalProduct>();
   }
}
boost::shared_ptr<EnvironmentalAbandonmentOfSchemeEnum> EnvironmentalPhysicalLeg::getAbandonmentOfScheme()
{
   if(!this->abandonmentOfSchemeIsNull_){
        if(abandonmentOfSchemeIDRef_ != ""){
             return boost::shared_static_cast<EnvironmentalAbandonmentOfSchemeEnum>(IDManager::instance().getID(abandonmentOfSchemeIDRef_));
        }else{
             return this->abandonmentOfScheme_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EnvironmentalAbandonmentOfSchemeEnum>();
   }
}
boost::shared_ptr<AdjustableOrRelativeDate> EnvironmentalPhysicalLeg::getDeliveryDate()
{
   if(!this->deliveryDateIsNull_){
        if(deliveryDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(deliveryDateIDRef_));
        }else{
             return this->deliveryDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
boost::shared_ptr<DateOffset> EnvironmentalPhysicalLeg::getPaymentDate()
{
   if(!this->paymentDateIsNull_){
        if(paymentDateIDRef_ != ""){
             return boost::shared_static_cast<DateOffset>(IDManager::instance().getID(paymentDateIDRef_));
        }else{
             return this->paymentDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<DateOffset>();
   }
}
boost::shared_ptr<BusinessCentersReference> EnvironmentalPhysicalLeg::getBusinessCentersReference()
{
   if(!this->businessCentersReferenceIsNull_){
        if(businessCentersReferenceIDRef_ != ""){
             return boost::shared_static_cast<BusinessCentersReference>(IDManager::instance().getID(businessCentersReferenceIDRef_));
        }else{
             return this->businessCentersReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessCentersReference>();
   }
}
boost::shared_ptr<BusinessCenters> EnvironmentalPhysicalLeg::getBusinessCenters()
{
   if(!this->businessCentersIsNull_){
        if(businessCentersIDRef_ != ""){
             return boost::shared_static_cast<BusinessCenters>(IDManager::instance().getID(businessCentersIDRef_));
        }else{
             return this->businessCenters_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessCenters>();
   }
}
boost::shared_ptr<XsdTypeBoolean> EnvironmentalPhysicalLeg::getFailureToDeliverApplicable()
{
   if(!this->failureToDeliverApplicableIsNull_){
        if(failureToDeliverApplicableIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(failureToDeliverApplicableIDRef_));
        }else{
             return this->failureToDeliverApplicable_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
boost::shared_ptr<EEPParameters> EnvironmentalPhysicalLeg::getEEPParameters()
{
   if(!this->eEPParametersIsNull_){
        if(eEPParametersIDRef_ != ""){
             return boost::shared_static_cast<EEPParameters>(IDManager::instance().getID(eEPParametersIDRef_));
        }else{
             return this->eEPParameters_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EEPParameters>();
   }
}
}

